cc/mcts_node.cc

```diff
@@ -0,0 +1,254 @@
+}
+
+std::string MctsNode::Describe() const {
+  using std::setprecision;
+  using std::setw;
+
+  auto child_action_score = CalculateChildActionScore();
+  using SortInfo = std::tuple<float, float, int>;
+  std::array<SortInfo, kNumMoves> sort_order;
+  for (int i = 0; i < kNumMoves; ++i) {
+    sort_order[i] = {child_N(i), child_action_score[i], i};
+  }
+  std::sort(sort_order.begin(), sort_order.end(), std::greater<SortInfo>());
+
+  std::ostringstream oss;
+  oss << std::fixed;
+  oss << setprecision(4) << Q() << "\n";
+  oss << MostVisitedPath() << "\n";
+  oss << "move : action     Q      U      P    P-Dir    N  soft-N  p-delta  "
+         "p-rel\n";
+
+  float child_N_sum = 0;
+  for (const auto& e : edges) {
+    child_N_sum += e.N;
+  }
+  for (int rank = 0; rank < 15; ++rank) {
+    Coord i = std::get<2>(sort_order[rank]);
+    if (child_N(i) == 0) {
+      break;
+    }
+    float soft_N = child_N(i) / child_N_sum;
+    float p_delta = soft_N - child_P(i);
+    float p_rel = p_delta / child_P(i);
+    // clang-format off
+    oss << std::left << setw(5) << i.ToKgs() << std::right
+        << ": " << setw(6) << setprecision(3) << child_action_score[i]
+        << ", " << setw(6) << child_Q(i)
+        << ", " << setw(5) << child_U(i)
+        << ", " << setw(5) << child_P(i)
+        << ", " << setw(5) << original_child_P(i)
+        << " " << setw(5) << static_cast<int>(child_N(i))
+        << " " << setw(5) << setprecision(4) << soft_N
+        << " " << setw(8) << setprecision(5) << p_delta
+        << " " << setw(5) << setprecision(2) << p_rel << "\n";
+    // clang-format on
+  }
+  return oss.str();
+}
+
+std::string MctsNode::MostVisitedPath() const {
+  std::ostringstream oss;
+
+  const auto* node = this;
+  while (!node->children.empty()) {
+    int next_kid = ArgMax(
+        node->edges,
+        [](const EdgeStats& a, const EdgeStats& b) { return a.N < b.N; });
+    auto it = node->children.find(next_kid);
+    if (it == node->children.end()) {
+      oss << "GAME END";
+      break;
+    }
+    node = it->second.get();
+    oss << node->move.ToKgs() << " (" << static_cast<int>(node->N())
+        << ") ==> ";
+  }
+  oss << std::setprecision(5) << "Q: " << node->Q();
+  return oss.str();
+}
+
+void MctsNode::InjectNoise(const std::array<float, kNumMoves>& noise) {
+  for (int i = 0; i < kNumMoves; ++i) {
+    edges[i].P = 0.75f * edges[i].P + 0.25f * noise[i];
+  }
+}
+
+MctsNode* MctsNode::SelectLeaf() {
+  auto* node = this;
+  for (;;) {
+    ++node->stats->N;
+
+    // If a node has never been evaluated, we have no basis to select a child.
+    if (node->state == State::kCollapsed) {
+      node->state = State::kSelected;
+      return node;
+    }
+    // If the node has already been selected for the next inference batch, we
+    // shouldn't select it again.
+    if (node->state == State::kSelected) {
+      // Revert the visits we just made.
+      --node->stats->N;
+      while (node != this) {
```

Comment:
is a do-while loop not kosher?

Reply:
I originally tried a do-while loop, but that doesn't handle the case where node == this. You need to break from inside the middle of the loop, which is awkward:
```C++
  while (true) {
    --node->stats->N;
    if (node == this) {
      break;
    }
    node = node->parent;
  }
```